File: go/general/139. 单词拆分.go
package main

func wordBreak(s string, wordDict []string) bool {
	n := len(s)
	dp := make([]bool, n+1)
	dp[0] = true
	for i := 1; i < n+1; i++ {
		dp[i] = false
	}
	for i := 0; i < n+1; i++ {
		for j := i + 1; j < n+1; j++ {
			if dp[i] && inString(s[i:j], wordDict) {
				dp[j] = true
			}
		}
	}
	return dp[n]
}

func inString(s string, t []string) bool {
	for _, v := range t {
		if s == v {
			return true
		}
	}
	return false
}

File: python/动态规划/139. 单词拆分.py
from typing import List

# dp[i]表示s的前i位是否可以用wordDict中的单词表示
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        n = len(s)
        dp = [False] * (n+1)
        dp[0] = True
        for i in range(n+1):
            for j in range(i+1,n+1):
                if(dp[i] and s[i:j] in wordDict):
                    dp[j] = True
        return dp[-1]
ans = Solution()
s = "applepenapple"
wordDict = ["apple", "pen"]
print(ans.wordBreak(s,wordDict))